Update Multi-cluster architecture doc

luolanzone opened the pull request:
Update MC architecture doc and change old diagrams' format from png to svg.

Signed-off-by: Lan Luo <luola@vmware.com>

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+need to choose the same tunnel type in a ClusterSet.
+
+To support network connectivity in a ClusterSet, Antrea Multi-cluster introduces two new CRDs `TunnelEndpoint` and
+`TunnelEndpointImport`, which helps to exchange member cluster's basic network information. It relys on existing
```
helps -> help

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+
+<img src="assets/mc-dataplane.png" width="600" alt="Antrea Multi-cluster Data Plane">
+
+Antrea agent is responsible to setup a tunnel between each member cluster's Gateway Node. Currently, Multi-cluster
```
set up

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+<img src="assets/mc-dataplane.png" width="600" alt="Antrea Multi-cluster Data Plane">
+
+Antrea agent is responsible to setup a tunnel between each member cluster's Gateway Node. Currently, Multi-cluster
+data plane will only work with `encap` mode. It will use the Geneve tunnel interface by default. Each member cluster
```
only works

jianjuns replied:
It uses

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+
+Antrea agent is responsible to setup a tunnel between each member cluster's Gateway Node. Currently, Multi-cluster
+data plane will only work with `encap` mode. It will use the Geneve tunnel interface by default. Each member cluster
+need to choose the same tunnel type in a ClusterSet.
```
needs

Could you check plural/singular in the whole doc?

jianjuns reviewed (commented):
I think we generally use only SVG diagram in the repo.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
 the resource export/import pipeline will ensure member clusters receive this ACNP spec to be replicated.
 Each member cluster's Multi-cluster Controller will then create an ACNP in their respective clusters.
+
+## Antrea Multi-cluster Data plane
```
Why call it data plane, but not "multi-cluster connectivity" or "cross-cluster connectivity", or even "multi-cluster tunnel"?

luolanzone replied:
I renamed it as `Antrea Multi-cluster Service Connectivity`, hope it's ok.

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+## Antrea Multi-cluster Data plane
+
+Antrea Multi-cluster support data plane since 1.7.0. It introduces a new feature gate in Antrea agent which
+is called `Multicluster`. This can help to support Service access between member clusters when Pod IPs in member
```
done, add a new section: `Antrea Multi-cluster Service Traffic Walk`

luolanzone replied:
done, add a new section `Antrea Multi-cluster Service Traffic Walk`

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -103,3 +103,23 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+
+## Antrea Multi-cluster Data plane
+
+Antrea Multi-cluster support data plane since 1.7.0. It introduces a new feature gate in Antrea agent which
```
rename it to `Service connectivity`

luolanzone commented:
Hi @jianjuns  I changed a little bit for MC doc, Could you review again? thanks.

jianjuns commented on docs/multicluster/user-guide.md:
```diff
@@ -142,10 +142,15 @@ We recommend creating one ServiceAccount for each member for fine-grained access
        namespace: antrea-mcs-ns
```
Could you fix the invalid link to "installation-in-a-shared-cluster" in "Deploy Mulit-cluster Controller"? BTW, the deployment guide looks complex. I think we should do a quick-start section first with the simplest deployment steps, before talking abut different options. But probably let us fix it later.

jianjuns commented on docs/multicluster/user-guide.md:
````diff
@@ -142,10 +142,15 @@ We recommend creating one ServiceAccount for each member for fine-grained access
    ```
 
-2. Copy the access token into the member cluster `test-cluster-east`. E.g.
+2. Generate the access token file from leader cluster, and apply it in the member cluster `test-cluster-east`, e.g.
````
the leader cluster

jianjuns commented on docs/multicluster/user-guide.md:
````diff
@@ -142,10 +142,15 @@ We recommend creating one ServiceAccount for each member for fine-grained access
 
    ```bash
+   # Generate the file 'member-east-access-token.yml' from your leader cluster
+   export KUBECONFIG=/path/to/kubeconfig-of-leader-cluster
````
Better not to export KUBECONFIG. Could you think about one of the following:

1. KUBECONFIG=/path/to/kubeconfig-of-leader-cluster && \
2. Use kubectl context
3. Use --kubeconfig

jianjuns replied:
Probably we should make fetching token a separate step.

jianjuns reviewed (commented):
Remove on-prem, public, etc. from the diagrams.

jianjuns commented on docs/multicluster/user-guide.md:
````diff
@@ -503,6 +508,102 @@ Events:
   Warning ACNPImportFailed     2m11s  resourceimport-controller  ACNP Tier random does not exist in the importing cluster test-cluster-west
 ```
 
+## Antrea Multi-cluster Deployment with Service Connectivity
````
I assume Gateway is a must for MCS since Antrea 1.7? In this case, this section should be moved to "Antrea Multi-cluster Installation".

jianjuns replied:
I think we need a topology diagram in this doc, which talks about three clusters forming a ClusterSet, with one to be both leader and member. And we should add Service CIDR and Gateway IPs for each cluster. And based on the information, we document the steps to create a ClusterSet with Gateways.

jianjuns commented on docs/multicluster/user-guide.md:
```diff
@@ -503,6 +508,102 @@ Events:
+  clusterID: test-cluster-east
+  gatewayInfos:
+  - gatewayIP: 10.17.27.55
+  serviceCIDR: 10.96.0.0/18
```
Do not use the default CIDR, as we do not support overlapping.

jianjuns commented:
@luolanzone : the current version of user guide looks too complex to follow. I think let us split the PR to two, one for architecture doc, one for user guide, and let me work together with you on the latter.

luolanzone committed:
Update Multi-cluster architecture doc

Update the Multi-cluster architecture doc with descriptions about
Multi-cluster Gateway, and change the diagrams' format from png to svg.

Signed-off-by: Lan Luo <luola@vmware.com>
Co-authored-by: Jianjun Shen <shenj@vmware.com>

luolanzone commented:
@jianjuns  I removed this PR from release v1.7, but could you help to review and check if it's Ok to merge in this release? thanks.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -57,7 +57,7 @@ and encapsulates them into ResourceImports.
 ## Service Export and Import
 
-<img src="assets/resource-export-import-pipeline.png" width="800" alt="Antrea Multi-cluster Resource Export/Import Pipeline">
+<img src="assets/resource-export-import-pipeline.svg" width="800" alt="Antrea Multi-cluster Resource Export/Import Pipeline">
```
The picture is a little small too (hard for me to read the texts).

jianjuns replied:
In the picture, consider:

1. Endpoint -> Endpoints
2. Monitor -> Watch?

luolanzone replied:
Done

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -106,3 +108,52 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+will be responsible to forward all cross-clusters traffic from local cluster to other member clusters.
+Below is the basic architecture for Antrea Multi-cluster Service network connectivity.
+
+<img src="assets/mc-service-connectivity.svg" width="600" alt="Antrea Multi-cluster Service Connectivity">
```
Again, a little small.

jianjuns replied:
We should remove "node-a1" around the tunnel line?

A minor one - the color of Cluster C box looks too different from that of Cluster A B. Could we adjust the colors?

luolanzone replied:
yea, node-a1 is added by mistake, it's fixed, and color is adjusted.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -106,3 +108,52 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
 the resource export/import pipeline will ensure member clusters receive this ACNP spec to be replicated.
 Each member cluster's Multi-cluster Controller will then create an ACNP in their respective clusters.
+
+## Antrea Multi-cluster Service Connectivity
```
I feel "Multi-cluster Service Connectivity" is not a good term for people to understand.

How about "Multi-cluster Gateway" or "Cross-cluster Connectivity".

luolanzone replied:
Rename it to Multi-cluster Gateway"

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -106,3 +108,52 @@ Multi-cluster admins can also specify certain ClusterNetworkPolicies to be repli
+
+## Antrea Multi-cluster Service Connectivity
+
+Antrea Multi-cluster supports Service connectivity directly since 1.7.0. It introduces a new feature gate
```
I feel no need to mention feature gate or particular component configurations in the architecture doc.

luolanzone replied:
Removed

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -84,14 +84,16 @@ been created by the Importer earlier.
 
```
The pipeline is introduced first considering both MC Service and MC network policy also depends on this, I will check how to refine it. thanks.

luolanzone replied:
Refined the whole structure and some descriptions.

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -8,7 +8,7 @@ ClusterSet, and enforced in all member clusters.
 The diagram below depicts a basic multi-cluster topology in Antrea.
 
-<img src="assets/basic-topology.png" width="500" alt="Antrea Multi-cluster Topology">
+<img src="assets/basic-topology.svg" width="500" alt="Antrea Multi-cluster Topology">
```
Updated.

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -8,7 +8,7 @@ ClusterSet, and enforced in all member clusters.
 
 The diagram below depicts a basic multi-cluster topology in Antrea.
```
Updated

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -8,7 +8,7 @@ ClusterSet, and enforced in all member clusters.
-<img src="assets/basic-topology.png" width="500" alt="Antrea Multi-cluster Topology">
+<img src="assets/basic-topology.svg" width="500" alt="Antrea Multi-cluster Topology">
 
 Given a set of Kubernetes clusters, there will be a leader cluster and several member clusters.
```
Done

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -8,7 +8,7 @@ ClusterSet, and enforced in all member clusters.
+<img src="assets/basic-topology.svg" width="500" alt="Antrea Multi-cluster Topology">
 
 Given a set of Kubernetes clusters, there will be a leader cluster and several member clusters.
 By default, a leader cluster itself is also a member cluster of a ClusterSet. A cluster
```
Done

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -32,35 +34,50 @@ interface for resource export/import that can be read/written by all member and
 
+## Antrea Multi-cluster Service
+
+Since Antrea v1.7.0, Multi-cluster Gateways must be configured to support Multi-cluster Service
```
This section makes no sense to be here. And readers do not even know what is Multi-cluster Service from this doc so far.

Anyway, let us do some small adjustment in this PR, and revise later. Let us move this section to be after "Antrea Multi-cluster Controller". And in this section, let us first have the "Service Export and Import" sub-section; and add another sub-section called "Service Access Across Clusters", and move this paragraph there.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -80,18 +97,53 @@ Service `cluster1-ns-foo-service`, `cluster2-ns-foo-service` and associated Endp
-any extra setting.
+## Antrea Multi-cluster Gateway
+
+Antrea Multi-cluster supports Service connectivity directly since 1.7.0. User can choose one of K8s Nodes
```
Antrea started to support Multi-cluster Gateway since v1.7.0.

luolanzone replied:
Done

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -6,13 +6,15 @@ ClusterSet. Antrea Multi-cluster also supports Antrea ClusterNetworkPolicy repli
+ClusterSet.
 
 ## Terminology
 
```
Fix the following:

> member of a ClusterSet with unique cluster ID, to claim a ClusterSet with unique ClusterSet ID.

with a unique cluster ID

and to claim a ClusterSet with a unique ClusterSet ID.

jianjuns replied:
It might be more natural to talk about ClusterSet first:

The ClusterClaim CRD is used to claim a ClusterSet with a unique ClusterSet ID, and to claim the cluster itself as a member of a ClusterSet with a unique cluster ID.

luolanzone replied:
Done

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -80,18 +97,53 @@ Service `cluster1-ns-foo-service`, `cluster2-ns-foo-service` and associated Endp
+cross-clusters traffic from local cluster to other member clusters. Below is the basic architecture for
+Antrea Multi-cluster connectivity with Gateway.
+
+<img src="assets/mc-gateway.svg" width="800" alt="Antrea Multi-cluster Gateway">
```
Yeah, they are the same font size, it's probably caused by different distance of boxes, I adjust them a little bit.

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -6,13 +6,15 @@ ClusterSet. Antrea Multi-cluster also supports Antrea ClusterNetworkPolicy repli
+and two member clusters.
 
-<img src="assets/basic-topology.png" width="500" alt="Antrea Multi-cluster Topology">
+<img src="assets/basic-topology.svg" width="650" alt="Antrea Multi-cluster Topology">
```
Done

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -6,13 +6,15 @@ ClusterSet. Antrea Multi-cluster also supports Antrea ClusterNetworkPolicy repli
-Given a set of Kubernetes clusters, there will be a leader cluster and several member clusters.
-By default, a leader cluster itself is also a member cluster of a ClusterSet. A cluster
-can also be configured as a dedicated leader cluster of multiple ClusterSets.
+An Antrea Multi-cluster ClusterSet includes a leader cluster and multiple member clusters.
```
Done

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,40 @@ given ClusterSet.
 and encapsulates them into ResourceImports.
 
-## Service Export and Import
+### Service Export and Import
```
I meant adding a section at the same level of "## Antrea Multi-cluster Controller" called "## Multi-cluster Controller Service", so the sections are like:

```
## Antrea Multi-cluster Controller
## Multi-cluster Controller Service
### Service Export and Import
### Service Access Across Clusters
## Multi-cluster Gateway
### Multi-cluster Service Traffic Walk
## Antrea Multi-cluster NetworkPolicy
```

What you think? BTW, I removed "Antrea" from a few section titles. I feel no need to repeat it in all places.

luolanzone replied:
sure, I will refine it. thanks.

luolanzone replied:
I suppose you mean `## Multi-cluster Service` in the second item?

jianjuns replied:
Ah, yes

luolanzone replied:
Done

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -6,13 +6,15 @@ ClusterSet. Antrea Multi-cluster also supports Antrea ClusterNetworkPolicy repli
 
-The diagram below depicts a basic multi-cluster topology in Antrea.
+An Antrea Multi-cluster ClusterSet includes a leader cluster and multiple member clusters.
+Antrea Multi-cluster Controller needs to be deployed in the leader and all member
```
Either this line or the next is not aligned at column 80?

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
-In a member cluster, Antrea Multi-cluster creates a Deployment that runs Antrea Multi-cluster
-Controller which is responsible for exporting resource to and importing resource from a leader
-cluster in a ClusterSet.
+Antrea Multi-cluster Controller watches ClusterSet, ClutserClaim, Service, ServiceExport etc.
```
Remove "watches...". It is not more important than other responsibilities of MC Controller to mention first.

jianjuns replied:
If you want, we can say "Antrea Multi-cluster Controller implements ClusterSet management and resource export/import in the ClusterSet".

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
-Controller which is responsible for exporting resource to and importing resource from a leader
-cluster in a ClusterSet.
+Antrea Multi-cluster Controller watches ClusterSet, ClutserClaim, Service, ServiceExport etc.
+It has different responsibilities in member and leader clusters.
```
In either a leader or a member cluster, Antrea Multi-cluster Controller is deployed with a Deployment of a single replica, but it takes different responsibilities in member and leader clusters.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
-In a leader cluster, Antrea Multi-cluster creates a Deployment that runs Antrea Multi-cluster
-Controller which is responsible for converting resources from different member clusters into one
-encapsulated resource as long as these resources have the same kind and match Namespace sameness.
+In a member cluster, Antrea Multi-cluster creates a Deployment with one replica that runs
```
I feel we can remove this paragraph, as resource export/import is talked in a later one too.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
+resource from a leader cluster in a ClusterSet.
 
-In ClusterSet initialization, Antrea Multi-cluster Controller in a member cluster watches
+In a leader cluster, Antrea Multi-cluster creates a Deployment with one replica that runs
```
Ditto

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
+member clusters into one encapsulated resource as long as these resources have the same kind
+and match Namespace sameness.
+
+When ClusterSet is initialized in member cluster, Antrea Multi-cluster Controller watches
```
Maybe add a sub-section of "ClusterSet Initialization"

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,33 +36,42 @@ given ClusterSet.
-and writes the ResourceExports to leader cluster. For resource importing, it watches ResourceImports
-from leader cluster, and creates multi-cluster Services and Endpoints with a prefix `antrea-mc-`
-plus exported Service name, and also ServiceImports which have the same name as the Service name.
+For resources exporting in the member cluster, Antrea Multi-cluster Controller watches ServiceExport,
```
This section is not well organized. I feel harder to explain all my suggestion than just writing down what is in my mind. Probably let me push a version for you to review.

jianjuns commented on docs/multicluster/architecture.md:
```diff
@@ -34,64 +37,152 @@ given ClusterSet.
+other member clusters through tunnels. The diagram below depicts Antrea
+Mulit-cluster connectivity with Multi-cluster Gateways.
+
+<img src="assets/mc-gateway.svg" width="800" alt="Antrea Multi-cluster Gateway">
```
Should we use "pod-a/b/c" to be consistent with "node-a1", etc.?

luolanzone replied:
sure, I will update it.

luolanzone replied:
@jianjuns  the change looks good to me overall, I changed a few minor descriptions and the line length for `Antrea Multi-cluster NetworkPolicy` section as well, you can move forward if you feel it's ready.

jianjuns reviewed (commented):
@luolanzone : I updated a version. Please review.

luolanzone commented on docs/multicluster/architecture.md:
```diff
@@ -34,75 +37,167 @@ given ClusterSet.
+leader cluster to join the ClusterSet.
+
+In the leader cluster, Multi-cluster controller watches and validates the
+ClusterSet and Clusterclaim CRs, and initializes the ClusterSet. It also validates
+the MemberClusterAnnounce CR created by a member cluster and updates the member
+cluster status to the ClusterSet.
```
Here is the main one I updated, we have a webhook for `MemberClusterAnnounce` in leader cluster, so I used `validates` here, and Admin is the one to add/update member clusters to ClusterSet, Controller will validate and update status only.

jianjuns committed:
Fixes of typos & formatting and a few minor revisions

jianjuns commented:
/skip-all
